### DISCUSSION TITLE
Position value returns nan after using ext_pos() to send position data getting from motion capture

### DISCUSSION THREAD

**Discussion by cc299792458** [bitcraze/discussions#654] [Q&A] [answered]
Hi, 

I meet a problem that the position value turns to be nan after I sending data getting from VICON(motion capture).
I use kalman estimator. I use ext_pos() function to send the data. I send x=0.0, y=0.0, z=0.0 to it continuosly.
I use LogConfig to log 'stateEstimate.x' and print it.

Before the data becomes nan, I find it first to be -2 or something like that in other trials, which is normal because the internal sensor can't estimate the position value right, then it becomes 0.999, which is also right. But in the next print, it becomes very huge and finally becomes nan.(see the figure) So I guess maybe it's because the wrong calculation of kalman filter or something like that?

Any advice would be appreciate!

![Screenshot from 2023-04-12 17-11-39](https://user-images.githubusercontent.com/52767591/231411317-c2e7103f-5182-419c-a5e4-5f433ea0c68e.png)

**Comment by cc299792458** [comment]
I think I can discuss the picute with more detials.

The "-1.6" in the pitcure is estimated by the internal sensors of crazyflie, at that time, it hasn't recieved the data sent by "ext_pos()".
Then, when it first recieved the data, it estimated the position to be near 0 in the next 4 rounds.
Finally, it becomes nan suddenly, at this time, it still recieved the data sent by ext_pos()

**Comment by krichardsson** [comment]
It looks like some computation in the kalman filter goes bad, something like a divide by 0 or similar. To me it looks like a bug, the state should never contain a NaN. The question is when/why it happens. 

You can turn on an extra check in the firmware, if you want, by enabling the line [`// #define DEBUG_STATE_CHECK` in `kalman_core.c`](https://github.com/bitcraze/crazyflie-firmware/blob/d183a2c0a73db4372eedb4106fb66581a8af852a/src/modules/src/kalman_core/kalman_core.c#L67). This will assert that the state is valid. Most likely the assert will trigger and your Crazyflie will reboot with a message in the console log pointing at the `assertStateNotNaN()` function. It will not solve the problem, but at least you will know :-)

The actual handling of the position can be found in `mm_position.c`, not sure what could go wrong there. 

You could try to change the standard deviation that is used for the position data that is pushed in, to see if this changes anything or provides any clues. You can set the std dev through the `locSrv.extPosStdDev` parameter.

If I understand correctly, you are not pushing in any actual data from the Vicon system, you are just pushing the position (0, 0, 0)? This sounds like a valid thing to do. 
If you believe your set up is good, I think we should treat this as a bug and convert this discussion into an issue instead. Let us know what you find out.

**Reply by cc299792458** [reply]
Hi sir,

I do a serious of experiments and I would like to share them with you. By the way, as you metioned, at this time, I didn't use the data of Vicon but just push (0, 0, 0) to crazyflie.

The first thing I did was enabling the line `// #define DEBUG_STATE_CHECK in kalman_core.c` and tried another time, it do rebooted. :-)

Then, I tried to modify the parameter `locSrv.extPosStdDev`, but I dit not set them by crazyflie-lib-python but modified them directly in the firmware, I think thes two ways are the same but the second one is more stable. I tried to set it to 100, 0.1, 0.01, 0.001, 0.000001 and observed the performance of crazyflie.

I know the default value is 0.01, but when I post this Q&A, at that time, I already set it to 0.001 because I know Vicon is capable to achieve that precision.

I first tried 0.000001, because I think if the `locSrv.extPosStdDev` is small enought, it means the estimator will nearly fully believe the external data, so it will be very near to the data(which is zero) I send to it. However, it failed. The performance is exact same to what I described former.

Then, I tried 100, because I think that's the opposite of a small dev, which means totally not to trust the external data. What's surprised, at this time, the crazyflie didn't reboot and the program seems capable to sustain. However, the estimated value is not accurate enough, I sent 0 as x to it, and the estimated value is about 2, and continously changed.

Insired by that, altoughd I didn't know the reason, I tried to set it back to 0.01. For this value, I test for 5 times. 4 times worked well and only 1 time rebooted. I carefully observe the phenomenon and I find, for the susseccful cases, the crazyflie recieved the data at an early time, which means the orignal estimated position x value is not far away from 0(like 1). And for the failed one, when the crazyflie recieved the data, the original extimated position x value is like 5, which is a little far away from 0. 
If this is confused, Let me explained with more details. When the crazyflie starts, the postion x will be set to 0. Because the internal sensor of crazyflie can't estimate the postion value well, these value will flow and continously increase. Like, after 5 seconds, the x value will be 2 and after 10 seconds, it will be 4.

After that, I do another trials with 0.1 and 1. I think the tolerance will increase but it's not obvious because they are also failed for a margin of 5.

Finally, I tried 100 again and I believe the tolerance increased.

That's all what I did. Maybe we can discuss together?

Thanks.

**Comment by krichardsson** [comment]
Thanks for your extensive tests!
I would like to try this out my self, I'll try to do it as soon as possible. 
I'll get back to you

**Reply by cc299792458** [reply]
Thanks, I got it!

**Reply by krichardsson** [reply]
I did try to reproduce this today, but I did not manage to get the same result. Could you run my test code on your system to see if you still get the problem? 

``` python 
import time

import cflib.crtp
from cflib.crazyflie import Crazyflie
from cflib.crazyflie.log import LogConfig
from cflib.crazyflie.syncCrazyflie import SyncCrazyflie
from cflib.crazyflie.syncLogger import SyncLogger
from cflib.utils import uri_helper

# URI to the Crazyflie to connect to
uri = uri_helper.uri_from_env(default='radio://0/80/2M/E7E7E7E7E7')

if __name__ == '__main__':
    cflib.crtp.init_drivers()

    lg_stab = LogConfig(name='state', period_in_ms=100)
    lg_stab.add_variable('stateEstimate.x', 'float')

    with SyncCrazyflie(uri, cf=Crazyflie(rw_cache='./cache')) as scf:
        cf = scf.cf

        cf.param.set_value('stabilizer.estimator', '2')
        cf.param.set_value('locSrv.extPosStdDev', 0.01)

        next_pos_update = time.time() + 2
        with SyncLogger(scf, lg_stab) as logger:
            for log_entry in logger:
                data = log_entry[1]
                print(data)

                if time.time() >= next_pos_update:
                    next_pos_update = next_pos_update + 0.1
                    cf.extpos.send_extpos(0.0, 0.0, 0.0)
```

**Reply by cc299792458** [reply]
That's a little weird, I tried your code and it will still reboot. Maybe the problem lies in the firmware?

**Reply by cc299792458** [reply]
Maybe you can try my firmware if you would like to. (https://github.com/cc299792458/marvel-firmware, choose branch: big-mode). But I'm sure, for kalman estimator, I did not change the code except the supurvisor of it.

**Reply by krichardsson** [reply, marked as answer]
I did a quick-ish test with your firmware, and to me it looks like something in the code is using too much CPU time, which is preventing other parts of the system to run as expected. Symptoms are for instance:
1. I sometimes get an assert in syslink, this usually happens when the system does not have enough time to empty incoming messages
2. The taks dump button in the console tab in the client is not working. The task dump runs as a timer and this could indicate that something else using a lot of CPU.

This might also be caused by some other change, but it seems as something is "a bit off" in the system.

I did take a quick look at your code and did a diff between when you started your branch and now. Unfortunately it looks like you copied in an old file tree at some point (Feb 18) which makes the diff hard to investigate as almost all files have been changed. I did not find the reason.

My suggestion is that you try to figure out if you have a task, timer or similar that uses a lot of CPU time, perhaps there is loop running all the time?
Either do it by reviewing your code, or (I find that this often is a good option) by starting from a new branch and re-write (copy/paste) from the old code tree.

**Reply by cc299792458** [reply]
Thank you for your suggestions! I decide to try yor second method(start from a new branch and re-write from the old code tree), By the way, I wonder, is there a way for me to find out if everything works well? So that I can do this test in every step.

**Reply by krichardsson** [reply]
A simple test would be to click the "Task dump" button in the console tab in the client. It should give you a dump of how much CPU each task has used. This functionality runs at the lowest priority level and if you get the task dump, you know this code has run and that there is still some slack in the system. 

The numbers in the dump shows how much CPU each task is using which might be useful for you to identify a task that consumes too much CPU time.

A note on the task dump: The counters are reset when you click the dump button. The first time you click it, it will show numbers from boot until now, which is maybe not very useful. The best way is to click it twice, with a few seconds in between, and look at the numbers in the second dump.

Good luck!

**Reply by cc299792458** [reply]
Thanks, I will try that!

**Reply by cc299792458** [reply]
Hi, sir. I think I find the reasons finally. Let me tell you this story (or you can just check the conclusion).

First of all, I tried the "Task dump" button with my customized firmware, but it seem to works well in my case.

Then, I tried to use ext_pos() with fresh firmware. It does work. What's important, I notice that when the external value was received, the kalman estimator was reset. That occurs to me that I have disabled this function. In my point of view, when the crazyflie gets the external position value for the first time, there is a big drop of the position value, and that cause a really big velocity, which may finally lead to some calculation in kalman estimator becomeing NaN. (check my attached figure to find the phenomenon) 

So I enable the reset of kalman estimator. And the ext_pos() works! I set the bound of velocity to 1000, but it still can be reset. That proves that there is an immense velocity value when it first recieves the external value.

But what's interesting, the ext_pose() still can't work with my firmware but can work with the fresh firmware. I, again, notice that there is a warning that says the prediction rate of kalman estimator is low. So I guess maybe I send the external value too fast. Originally, I use a loop with no delay to call the ext_pose() function. After I did my inference, I set the frequency of calling the function to 200Hz. I also test 100Hz and 500Hz. 100Hz can work well while 500Hz seems still too high.

The conclusion is: First, make sure the kalman estimator is able to reset. Second, send the command in a proper frequency(100Hz or 200Hz works fine).

That's all. It's really interesting to debug following your instructions. Although the process is tough and confusing sometimes.(I mean the problem is hard itself) :-)

![Screenshot from 2023-04-22 14-55-59](https://user-images.githubusercontent.com/52767591/233773132-328d69c3-6c8f-4ccb-9469-acc23f91be17.png)

**Reply by krichardsson** [reply]
Interesting that the estimator resets! Maybe it makes sense that the velocity becomes very high when you "force" it to a (0, 0, 0)? Not sure?

I usually push exp pos data at pretty low rates, like 5-10 Hz, but it depends on the situation. It makes sense that it use more CPU when pushing more, both the kalman estimator and the radio link.

Good that you found the problem!
